### DISCUSSION TITLE
How to create subdirectories for each disc for multi-disc albums?

### DISCUSSION THREAD

**Discussion by ali-ramadhan** [MoeMusic/Moe#229] [Q&A] [answered]
Kinda following from #226 I am trying to create a subdirectory for each disc for multi-disc albums but I am not sure how to do this.

Using an f-string in `track_path` cannot work since the f-strings cannot contain the `/` character (#226).

So I tried writing a plugin to do this (https://github.com/ali-ramadhan/moe-plugins/blob/d4df04ec438b763829a1cf588f0fbf1ff62d44c1/disc_and_track.py) which returns strings like `Disc 2/02`. But the `/` character gets converted into a `_` as expected since most filesystems do not allow `/` in file names, so you get file names like `Disc 2_02. Some track.flac`.

**Comment by jtpavlock** [comment]
Hmm, I think I see the issue more clearly now. So our current process for evaluating path templates is:

1. Split on "/" into path "parts"
2. Evaluate each part's f-string and functions
3. Sanitize those evaluations
4. Combine parts

You want a way to introduce new path parts on step 2. The functionality you're asking for is not currently supported, but it is something I would like to support and I think is a reasonable request, I just need to figure out the best way to do so.

One option that comes to mind would be to create a hook that allows plugins to edit the path template prior to it being split into parts. This hook would then return an f-string with programmatically placed `/` to later be split and sanitized.

Does that seem like it would work for you? Or do you have any other suggestions? I'm open to changing the process entirely if it makes sense to do so.

**Reply by ali-ramadhan** [reply]
Thanks for your feedback!

I'm also not sure of the best approach. I initially expected the process to have steps 1 and 2 switched. But I think the issue then is what happens if the artist's name or the album title have slashes in them.

> One option that comes to mind would be to create a hook that allows plugins to edit the path template prior to it being split into parts. This hook would then return an f-string with programmatically placed `/` to later be split and sanitized.

I think this makes sense, and it's especially cool that adding one extra hook adds this extra flexibility without complicating the code further.

I assume `track_path` remains unchanged, I would just have to do something like this?

```python
@moe.hookimpl
def create_path_template_func_pre_path_splitting():
    return [disc_and_track]
```

**Reply by jtpavlock** [reply]
> But I think the issue then is what happens if the artist's name or the album title have slashes in them.

Exactly, there wouldn't be an easy way to see which `/` are part of an album title and which are supposed to be path delimiters. 



> I assume `track_path` remains unchanged, I would just have to do something like this?

So I'd probably create a hook for each of tracks, albums, and extras called something like `override_album_path_template` which returns a string. This string would then override the path template configuration setting, but you'd still be able to refer to your custom functions and such.

**Reply by ali-ramadhan** [reply]
Ah interesting. I was thinking of something more like hooks that are evaluated before and after splitting on `/`. Then you can write something like

```toml
track_path = "{disc_and_track(track, album)}. {fancy_track_name(track)}{track.path.suffix}"
```

where `disc_and_track` is evaluated prior to `/` splitting so it can create subdirectories while `fancy_track_name` is evaluated after `/` splitting so the track name can be properly sanitized.

I think what you're suggesting is a hook that takes the almost-final `track_path` before `/` splitting and modifies it? Maybe I'm misunderstanding how this hook works as it feels like it might be more difficult to work with? I guess if you just want to add disc subdirectories it's not so bad as you just prefix `track_path` with a string, but more complex stuff like replacing the track name would require string parsing and regex substitutions?

**Comment by jtpavlock** [comment, marked as answer]
Going back to your original question, does something like this do what you want?

```python
def format_disc_dir(track, album):
    if album.disc_total == 1:
        return ""
    elif album.disc_total <= 9:
        return "Disc {track.disc:1d}"
    elif album.disc_total <= 99:
        return "Disc {track.disc:02}"
    elif album.disc_total <= 999:
        return "Disc {track.disc:03d}"
```

then your track path configuration would be:
`"{disc_and_track(track, album)}/{track.title}"`

If not, I'm a bit confused what you want, because the default path already splits tracks into separate disc directories.

**Reply by jtpavlock** [reply]
I'm actually very confused now that I look at your code a bit longer... would you mind giving a few examples on how exactly you mean to format your albums?

**Reply by ali-ramadhan** [reply]
Ah hmmm maybe that would work? I will try but then we would have a leading `/` in `track_path` for single disc albums I think?

I was thinking of create a directory structure like this for single disc albums:

```
Album with one disc
├── 01. Some track.flac
├── ...
├── cover.jpg
└── Some extra.txt
```

and a directory structure like this for multi-disc albums:

```
Album with multiple discs
├── cover.jpg
├── Disc 1
│  ├── 1-01. Some track.flac
│  ├── ...
│  └── Some disc specific extra.txt
└── Disc 2
   ├── 2-01. Some other track.flac
   ├── ...
   └── Some other disc specific extra.txt
```

Organizing the extras will require a separate plugin though.

**Reply by jtpavlock** [reply]
Yeah the default `"{f'Disc {track.disc:02}' if album.disc_total > 1 else ''}/"` should do exactly what you want. Double `//` will get squished into a single `/` so there won't be an empty directory

**Reply by jtpavlock** [reply]
And if you don't want leading zeroes remove the `:02` from the end of `track.disc`

**Reply by ali-ramadhan** [reply]
Ah nice! Didn't realize that double `//` would get squished, that's good to know. And yeah I only want leading zeros when `disc_total >= 10` which I think I will still get with your suggestion.

Thanks for your help! Now to see if I can get extras organized nicely.

**Reply by jtpavlock** [reply]
No problem, sorry to lead you down a rabbit hole before thinking through your intent a bit more. At least now I have an idea on how to implement more flexibility into path formats if the need ever arises.

**Reply by ali-ramadhan** [reply]
If I didn't rush to change the configuration file I might have realized the default did just what I wanted haha 🙃
